Lib/Protocols/IdFTPServer.pas

```diff
@@ -6018,6 +6060,77 @@ function TIdFTPServer.MLSFEATLine(const AFactMask: TIdMLSDAttrs;
   end;
 end;
 
+function ParseCSIDParams(const AParams : String) : string;
+var LBuf,
+    LValueValue : String;
+    LValueName : String;
+    LCltName : String;
+    LCltVersion : String;
+    LCltOS : String;
+begin
+   Result := '';
+   LCltName := '';
+   LCltVersion := '';
+   LBuf := AParams;
+   repeat
+      LValueValue := Fetch(LBuf,';');
+      LValueName := TrimLeft(Fetch(LValueValue,'='));
+      case PosInStrArray(LValueName,['Name','Version','OS']) of
+        0 : LCltName := LValueValue;
+        1 : LCltVersion := LValueValue;
+        2 : LCltOS := LValueValue;
+      end;
+   until LBuf = '';
+   Result := Trim(LCltName + ' ' + LCltVersion);
+   if LCltVersion <> '' then begin
+     Result := Result + ' ' + LCltOS;
+   end;
+end;
+
+procedure TIdFTPServer.CommandCSID(ASender: TIdCommand);
+var LClientInfo : String;
+   LServerInfo : String;
+   i : Integer;
+begin
+  if Assigned(FOnClientID) then begin
+    LClientInfo := ParseCSIDParams(ASender.UnparsedParams);
+    FOnClientID(ASender.Context as TIdFTPServerContext, LClientInfo);
+  end;
+  if FServerInfo.ServerName <> '' then begin

```

Comment:
If `ServerName` is empty but any of the other facts are not, then the reply will be malformed.  The separating `;` tokens need to be added dynamically based on whether `LServerInfo` is blank or not.